### ISSUE TITLE
Missed sink in a list initialised with a function

### ISSUE THREAD

**Issue by draftyfrog** [facebook/pyre-check#1002]
# Pysa Bug
**Pre-submission checklist**
[x] I've checked the list of [common issues](https://pyre-check.org/docs/pysa-quickstart/#common-issues) and mine does not appear

**Bug description**
I am running Pysa on the following code:
```python
class MyClass:
  def my_function(self):
    lst = MyClass.init([]) 
    lst.append(self.source()) # add tainted element to the list
    self.sink(lst[0]) # NOT reported by Pysa
  
  def source(self): 
    return "Secret" 
  
  def sink(self, param: str):
    return param

  @staticmethod
  def init(listi: list[str]):
    return listi 
```
The sink `self.sink(lst[0])` is not reported by Pysa, even though the argument is tainted. 

In my sources_sinks.pysa I declare `source` and `sink` as taint sources and sinks respectively:
```
def test.MyClass.source() -> TaintSource[TestSource]: ...
def test.MyClass.sink(param: TaintSink[TestSink]): ...
```

**Reproduction steps**
I ran Pysa version **0.9.25** using the command `pyre analyze --save-results-to ./results/ --infer-self-tito test.py`.

**Expected behavior**
Pysa should detect the issue and report`self.sink(lst[0])`.

**Additional context**
This seems related to the type checking issue described [here](https://github.com/facebook/pyre-check/issues/964). If I add a type annotation to `lst` in line 3 `lst:list[str] = MyClass.init([])`, Pysa correctly detects the issue.

**Comment by arthaud** [comment]
Hi @draftyfrog,

I don't think that'sexactly  the same problem.
What happens if you add a return type annotation to `init`:
```
  @staticmethod
  def init(listi: list[str]) -> list[str]:
    return listi 
```
I would expect Pysa to find the issue.
In general, Pysa works better with fully annotated code: functions should have argument annotations and return type annotations. In theory, local variables don't need to be annotated (but this can help).
I think running pyre in strict mode should tell you all functions missing type information.

**Comment by arthaud** [comment]
> Why does the type annotation makes a difference here, is it because Pysa does not understand that append() should propagate the taint in this case?

Without the type annotation, pyre/pysa doesn't know the type of `lst`. When it sees `lst.append(self.source())`, it doesn't know which function is called, it uses the obscure heuristic: https://www.internalfb.com/intern/staticdocs/pyre/docs/pysa-advanced/#obscure-models
By default, for unknown calls, we assume that taint on argument is returned, but we don't assume that taint is assigned to `self`, so this leads to the false negative.

> Thanks! Do you know when this would happen roughly?

Hopefully within the next 6 months, but I can't promise anything.
